### PR TITLE
[Merged by Bors] - Remove `NIPoST` and `NodeID` from types.ActivationTx

### PR DESCRIPTION
**Pull request by poszu**
## Motivation

Trimming down types.ActivationTx from unwanted fields. Prerequisite for ATX V2.

## Description

Removed:
- `NIPost`
- `NodeID`

## Test Plan

Existing tests pass

## TODO

- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [ ] Update documentation as needed
- [ ] Update [changelog](../CHANGELOG.md) as needed

### CODE REVIEW

File: activation/activation_test.go [hunk 1]
```diff
@@ -1579,28 +1547,21 @@ func TestGetPositioningAtxPicksAtxWithValidChain(t *testing.T) {
...
-		PositioningATX: tab.goldenATXID,
-		CommitmentATX:  &tab.goldenATXID,
-	}
-	nipostData := newNIPostWithPoet(t, []byte("0"))
-	invalidAtx := newAtx(ch, nipostData.NIPost, 2, types.Address{})
-	require.NoError(t, SignAndFinalizeAtx(sigInvalid, invalidAtx))
-	vInvalidAtx, err := invalidAtx.Verify(0, 100)
+	invalidAtx := newInitialATXv1(t, tab.goldenATXID)
+	invalidAtx.Sign(sigInvalid)
+	vInvalidAtx := toVerifiedAtx(t, invalidAtx)
```
Comment (fasmat) [inline activation/activation_test.go:1552]: Do we still need the `types.VerifiedActivationTx` type? I think it should not be needed any more after we have the domain `types.ActivationTx` type. Wdyt?

File: activation/validation_test.go [hunk 2]
```diff
@@ -541,60 +541,31 @@ func TestVerifyChainDeps(t *testing.T) {
...
-		vAtx.SetValidity(types.Unknown)
+		atx := newChainedActivationTxV1(t, goldenATXID, invalidAtx, goldenATXID)
+		atx.Sign(signer)
+		vAtx := toVerifiedAtx(t, atx)
 		require.NoError(t, atxs.Add(db, vAtx))
 
 		ctrl := gomock.NewController(t)
 		v := NewMockPostVerifier(ctrl)
-		v.EXPECT().Verify(ctx, (*shared.Proof)(atx.NIPost.Post), gomock.Any(), gomock.Any())
+		v.EXPECT().Verify(ctx, gomock.Any(), gomock.Any(), gomock.Any())
```
Comment (fasmat) [inline activation/validation_test.go:558]: Why the change? It should still verify the `Post` that was part of the atx or not?

```suggestion
		v.EXPECT().Verify(ctx, (*shared.Proof)(atx.NIPost.Post), gomock.Any(), gomock.Any())
```

File: checkpoint/recovery_test.go [hunk 3]
```diff
@@ -286,7 +286,7 @@ func validateAndPreserveData(
 
 		mvalidator.EXPECT().PositioningAtx(vatx.PositioningATX, cdb, goldenAtx, vatx.PublishEpoch)
 		mvalidator.EXPECT().
-			NIPost(gomock.Any(), vatx.SmesherID, gomock.Any(), vatx.NIPost, gomock.Any(), vatx.NumUnits, gomock.Any()).
+			NIPost(gomock.Any(), vatx.SmesherID, gomock.Any(), gomock.Any(), gomock.Any(), vatx.NumUnits, gomock.Any()).
```
Comment (fasmat) [inline checkpoint/recovery_test.go:289]: Same here, why make the assertion less strict?

```suggestion
			NIPost(gomock.Any(), vatx.SmesherID, gomock.Any(), vatx.NIPost, gomock.Any(), vatx.NumUnits, gomock.Any()).
```

**Review by fasmat — APPROVED** [review]

File: activation/activation_test.go (hunk 1 above)
Comment (poszu) [reply activation/activation_test.go:1552]: No, we shouldn't need it. Removing it is somewhere on the list TODO...

File: sql/atxs/atxs_test.go [hunk 4]
```diff
@@ -597,10 +597,7 @@ func TestLoadBlob(t *testing.T) {
 	require.Equal(t, []int{len(blob1.Bytes)}, blobSizes)
 
 	var blob2 sql.Blob
-	atx2, err := newAtx(sig, func(atx *types.ActivationTx) {
-		nodeID := types.RandomNodeID()
-		atx.NodeID = &nodeID // ensure ATXs differ in size
-	})
+	atx2, err := newAtx(sig)
```
Comment (ivan4th) [inline sql/atxs/atxs_test.go:601]: Is there a way to have different sizes in these ATXs or is the new wire type of constant size?
The test verifies that there's no mistake in the query that could be causing it to e.g. return the same size for different blobs

File: activation/validation_test.go (hunk 2 above)
Comment (poszu) [reply activation/validation_test.go:558]: The `atx` doesn't contain `Post`, only the wire type does. Keeping this arg-expectation would require translating from the wire type to `shared.Proof` which I figured is not worth the hassle.

File: sql/atxs/atxs_test.go (hunk 4 above)
Comment (poszu) [reply sql/atxs/atxs_test.go:601]: I don't understand, the test passes without this - is it missing some assertions? Perhaps you could clarify what's the purpose of this test, wdym by "return the same size for different blobs"?

Comment (fasmat) [reply sql/atxs/atxs_test.go:601]: @ivan4th an ATX can vary in size (i.e. have different number of bytes in it's blob). The reason for this is that it

a) initial ATXs contain 2 posts instead of one
b) initial ATXs (and occasionally non-initial ones) contain a post VRF
c) a PoST has a varying size, i.e. a post for 256 GiB is smaller than a PoST for 4 TiB

Comment (ivan4th) [reply sql/atxs/atxs_test.go:601]: There's `atxs.GetBlobSizes` func that returns the sizes of multiple blobs: https://github.com/spacemeshos/go-spacemesh/blob/213a58503544b24684acdd29fbffd6376876b996/sql/atxs/atxs.go#L334
which is based on `sql.GetBlobSizes`: https://github.com/spacemeshos/go-spacemesh/blob/213a58503544b24684acdd29fbffd6376876b996/sql/database.go#L527
It seems to me that it's relatively easy to imagine a potential bug in this func (which could be introduced by a future refactor) making it return wrong sizes for the blobs, e.g. same size for each one. Thus, I intentionally used ATXs with different sizes of the serialized representation to check it, as was noted in the comment in the replaced part of code

File: activation/validation_test.go (hunk 2 above)
Comment (fasmat) [reply activation/validation_test.go:558]: but the `atx` in this test is a `wire.ActivationTx1`, so it contains a post and the conversion happens in the actual implementation anyway. A proper conversion is also straight forward (if you want to avoid the implicit cast that might break at some point in the future):

```suggestion
		post := &shared.Proof{
			Pow:     atx.NIPost.Post.Pow,
			Indices: atx.NIPost.Post.Indices,
			Nonce:   atx.NIPost.Post.Nonce,
		}
		v.EXPECT().Verify(ctx, post, gomock.Any(), gomock.Any())
```

File: sql/atxs/atxs_test.go (hunk 4 above)
Comment (fasmat) [reply sql/atxs/atxs_test.go:601]: OK; I misunderstood your earlier comment. Yes the test should add ATXs of different sizes and now that I looked at the test closer I noticed a different problem @poszu 

This tests generates a `types.VerifiedActivationTx` that is then signed using `activation.SignAndFinalizeAtx` which also converts the domain type to a wire type, which isn't possible any more - the domain type doesn't contain a post any more.

This results in the blob being incomplete.

In my opinion `atxs.Add()` should take 2 arguments - the domain object and a `[]byte` that will be stored as the blob for the ATX with the ID in the domain type. This will probably require quite a few other tests to be updated (but only the `activation.Handler` in non-test code).

**Review by fasmat — CHANGES_REQUESTED** [review]
With the the PR change converting domain -> wire isn't actually possible any more and code that uses `wire.ActivationTxToWireV1` need to be updated to not require the conversion any more.

This will probably also require the update of the `sql/atx` functions that add ATXs to the db to receive both the domain type and the encoded ATX (in bytes) as argument.

File: activation/validation_test.go (hunk 2 above)
Comment (poszu) [reply activation/validation_test.go:558]: Yes, as I said above.

File: sql/atxs/atxs_test.go (hunk 4 above)
Comment (poszu) [reply sql/atxs/atxs_test.go:601]: @ivan4th Thanks for the explanation. I will fix the test and add the missing assertion verifying that the blobs in fact have different size.

> In my opinion atxs.Add() should take 2 arguments - the domain object and a []byte that will be stored as the blob for the ATX with the ID in the domain type. 

It already "takes" the blob which is part of the `types.VerifiedActivationTx`.

**Comment by poszu** [comment]
> With the the PR change converting domain -> wire isn't actually possible any more and code that uses `wire.ActivationTxToWireV1` need to be updated to not require the conversion any more.

Yes, the conversion from the domain to wire type is no longer sound. However, this only happens in some unit tests, which don't care about the missing data at all. At the end of the refactor conversion in this direction **will not** be present anywhere in the code. For the sake of keeping the PRs small and being able to continue with refactoring at a meaningful speed, I'm willing to accept this temporary unsoundness.

> This will probably also require the update of the `sql/atx` functions that add ATXs to the db to receive both the domain type and the encoded ATX (in bytes) as argument.

It's already done in the #5868.

File: sql/atxs/atxs_test.go (hunk 4 above)
Comment (fasmat) [reply sql/atxs/atxs_test.go:601]: > It already "takes" the blob which is part of the `types.VerifiedActivationTx.`

Yes, I remember we already had a similar discussion on a previous PR 😅

**Review by fasmat — APPROVED** [review]

**Review by ivan4th — APPROVED** [review]

**Comment by poszu** [comment]
bors merge
